### PR TITLE
feat(email): add option to limit number of EmailAddress on an account

### LINKED ISSUE #2621: Feature request: Allow choice between using multiple email addresses and a single email address
Not all websites **need** multiple email addresses and this might prove to be overwhelming for some coupled with the user experience in having multiple email addresses. 
I think a setting should exists such as `ACCOUNT_USE_MULTIPLE_EMAILS` with a default value of true which allows websites choose between working with one email address and working with the default multiple email addresses that this package presently provides. 
I have a piece of hacky, untested code that provides this feature together with a patch for an [issue](https://github.com/pennersr/django-allauth/issues/2620) I raised that could solve this problem. 
As stated in the issue, the code is still in its hacky stages and would be improved upon by the community.

### PR DESCRIPTION
**Pull request by iarp**
Fixes #2621

General

 - [x] Make sure you use [semantic commit messages](https://seesparkbox.com/foundry/semantic_commit_messages).
       Examples: `"fix(google): Fixed foobar bug"`, `"feat(accounts): Added foobar feature"`.
 - [x] All Python code must be 100% pep8 and isort clean.
 - [ ] JavaScript code should adhere to [StandardJS](https://standardjs.com).
 - [ ] If your changes are significant, please update `ChangeLog.rst`.
 - [ ] Feel free to add yourself to `AUTHORS`.

### CODE REVIEW

**Comment by 9mido** [comment]
You are on a roll @iarp! Keep it up! You can make 2 other quick PRs (hopefully they are quick) if you fix the 2 email spamming issues I commented a solution to which are also pretty bad and need fixes.

**Comment by stfenjobs** [comment]
please merge this pull  @pennersr 

thank you very much

File: ChangeLog.rst [hunk 1]
```diff
@@ -7,6 +7,9 @@ Note worthy changes
 - If ``ACCOUNT_LOGIN_ATTEMPTS_LIMIT`` is set and the user successfully
   resets their password, the timeout is cleared to allow immediate login.
 
+- New setting ``ACCOUNT_EMAIL_LIMIT_ON_ACCOUNT`` lets you decide how many
```
Comment (pennersr) [inline ChangeLog.rst:10]: How about `ACCOUNT_MAX_EMAIL_ADDRESSES` ?

File: allauth/account/forms.py [hunk 2]
```diff
@@ -432,6 +432,8 @@ def clean_email(self):
                               " with this account."),
             "different_account": _("This e-mail address is already associated"
                                    " with another account."),
+            'over_account_limit': _("You cannot add more email addresses as "
+                                    "you are at maximum allowance.")
```
Comment (pennersr) [inline allauth/account/forms.py:436]: Would be nice if this message simply said: "You cannot setup more than %d e-mail addresses.".

File: allauth/account/forms.py [hunk 3]
```diff
@@ -441,8 +443,15 @@ def clean_email(self):
             raise forms.ValidationError(errors["this_account"])
         if on_diff_account and app_settings.UNIQUE_EMAIL:
             raise forms.ValidationError(errors["different_account"])
+        if self.is_over_account_email_limit():
+            raise forms.ValidationError(errors['over_account_limit'])
         return value
 
+    def is_over_account_email_limit(self):
+        email_limit = app_settings.EMAIL_LIMIT_ON_ACCOUNT
+        users_emails = self.user.emailaddress_set.count()
```
Comment (pennersr) [inline allauth/account/forms.py:452]: You could skip doing this DB query if there is no limit configured.

File: allauth/templates/account/email.html [hunk 4]
```diff
@@ -44,14 +44,19 @@ <h1>{% trans "E-mail Addresses" %}</h1>
 
 {% endif %}
 
-
+  {% if can_add_emails %}
     <h2>{% trans "Add E-mail Address" %}</h2>
 
+    {% if account_email_limit and account_email_limit > 1 %}
```
Comment (pennersr) [inline allauth/templates/account/email.html:50]: From a less-is-more perspective, I think it is more user friendly to not constantly present this limit to the user. Likely, under normal circumstances, users should not be hitting the limit anyway... so I would drop this.

File: docs/configuration.rst [hunk 5]
```diff
@@ -89,6 +89,16 @@ ACCOUNT_EMAIL_MAX_LENGTH(=254)
   3-byte and 4-byte Unicode character sets
   <https://dev.mysql.com/doc/refman/5.5/en/charset-unicode-conversion.html>`_.
 
+ACCOUNT_EMAIL_LIMIT_ON_ACCOUNT(=None)
+  Limit how many EmailAddress records a user can add to their account.
+
+  Will not affect users that already have more than the limit after changing
+  this value later on in your projects lifespan.
+
+  Warning: An account must have at least One email address. A user cannot
```
Comment (pennersr) [inline docs/configuration.rst:98]: "One" caps.

File: docs/configuration.rst [hunk 6]
```diff
@@ -89,6 +89,16 @@ ACCOUNT_EMAIL_MAX_LENGTH(=254)
...
+  remove the only email address on their account. If you set this option to 1
+  they will be unable to ever add or change their current email address.
```
Comment (pennersr) [inline docs/configuration.rst:100]: This is true -- so shouldn't we just raise ImproperlyConfigured when the limit is set to 1 ?

**Review by pennersr — CHANGES_REQUESTED** [review]

File: docs/configuration.rst (hunk 6 above)
Comment (iarp) [reply docs/configuration.rst:100]: I'm not sure where it would be best to raise the error, in AppSettings init?

Comment (pennersr) [reply docs/configuration.rst:100]: You are already asserting that it is > 0 there -- why not make that > 1 ?

**Comment by 9mido** [comment]
I would like to add that personally I would want MAX_EMAIL_ADDRESSES=1 to prevent a user from adding an unverified email address that is not their own then spamming that other person's inbox by spam clicking the "re-send verification" button. This helps with CAN-SPAM and prevents the spam victim from saying "why did I receive this email from this website without opting in"? As for others who want more than 1 email per account plus all the other possible actions, the other timeout setting would be needed anyways just as an abuse prevention mechanism.

**Comment by pennersr** [comment]
Closed via 14c52ed1 -- thanks!

**Comment by 9mido** [comment]
Yay! Thank you :)

**Comment by 9mido** [comment]
@iarp I just tried testing this new setting and I have the following in my settings.py file:

ACCOUNT_MAX_EMAIL_ADDRESSES=1

In requirements.txt I have: 

git+https://github.com/pennersr/django-allauth

But on /accounts/email/ I can still add as many unverified emails as I want as if the setting does not work. 

Am I doing something wrong here?

**Comment by 9mido** [comment]
@pennersr 

I see in that PR you have - "You can now limit the amount of email addresses a user can associate to his
  account by setting ``ACCOUNT_MAX_EMAIL_ADDRESSES``.

But when I look at @iarp's code I see MAX_EMAIL_ADDRESSES.

When I do:

pip3 install git+https://github.com/pennersr/django-allauth

and in my settings file:

MAX_EMAIL_ADDRESSES=1

or 

ACCOUNT_MAX_EMAIL_ADDRESSES=1

on the /accounts/email/ page, I am able to add as many unverified emails as I want. It does not seem to be working for me. Am I doing something wrong here? 

Please let me know.

**Comment by iarp** [comment]
Working fine for me. Confirmed to be `ACCOUNT_MAX_EMAIL_ADDRESSES`.

**Comment by 9mido** [comment]
@iarp thank you for the confirmation. 

I did a fresh restart of my live linux USB and installed all my libraries from scratch from my wiped system. 

Not sure if the issue was upgrading from the version on PyPi to the new one using pip3 install git+https://github.com/pennersr/django-allauth command.

It does indeed work after I did installed from scratch on a new system. 

As always, and once again, thank you!

**Pull request closed without merging**